ArduinoHallEffect_good_code.ino

```diff
@@ -1,71 +1,62 @@
 //This code is full of test strings and is generally a conglomerate of code from stackexchange and
-//bad form. If it looks stupid, I probably wrote it. 
+//bad form. If it looks stupid, I probably wrote it.
 //Wesley Kagan, 2020
 
+//this might be of interest https://forum.arduino.cc/index.php?topic=4324.0 fast pin switching at 2MHz on Uno
+
 const int hall = 2;
+const int INTAKE_V = 13;
+const int EXHAUST_V = 12;
 int hallcounter = 0;
-int hallstate = 0;
-int lasthallstate = 0;
-unsigned long triggerTime;
-unsigned long last_triggerTime;
 unsigned long timeGap;
 unsigned long last_timeGap;
-unsigned int degree;
-unsigned int state;
-int newstate;
-
+int degree;
+int cycle = 1; //1 is intake cycle, -1 is exhaust, change initial cycle here
 
 void setup() {
-  pinMode(hall, INPUT);
-  Serial.begin(115200);
-  pinMode(13, OUTPUT);    // sets the digital pin 13 as output
-  pinMode(12, OUTPUT);    // sets the digital pin 12 as output
-  attachInterrupt(0, magnet_detect, RISING);//Initialize the intterrupt pin digital pin 2
-  degree = 0;
+   pinMode(hall, INPUT);
+   Serial.begin(115200);
+   pinMode(INTAKE_V, OUTPUT);                 // sets the digital pin 13 as output
+   pinMode(EXHAUST_V, OUTPUT);                // sets the digital pin 12 as output
+   attachInterrupt(0, magnet_detect, RISING); //Initialize the intterrupt pin digital pin 2
+   degree = 0;
 }
 
- void loop()
- {
-  //The compiler says this is needed otherwise the govt. takes my cat
-   }
+void loop() {
+   //The compiler says this is needed otherwise the govt. takes my cat
+}
 
 void magnet_detect() {
-      hallcounter ++;
-      triggerTime = millis();
-      timeGap = triggerTime - last_triggerTime;
-      last_triggerTime = triggerTime;
-     
-      if (timeGap >= last_timeGap + last_timeGap / 2)
-      {
-        //Serial.println(state);
-        Serial.println("missing tooth");
-        hallcounter = 1; 
-        state++; //This right here is garbage and you know it.
+   hallcounter++;
+   timeGap = millis() - timeGap;
+
+   if (timeGap >= (int)(1.5 * last_timeGap)) {

```

Comment:
Not sure I understand why this cast to an (int)?

Reply:
Probably to avoid a casting for timeGap to a float. Only numbers of the same type can be compared.

Reply:
Ah, thanks. It should probably be a (long) to keep it consistent with the type declaration of these variables.

Reply:
Consider multiplying the timeGap by 10 and last_timeGap by 15 to prevent calling into the floating point libraries. See this: https://github.com/nkapron/Freevalve_Arduino/blob/054797a65dec251340a225ef7532224851ff92c9/valve_control.ino#L142

Reply:
> 
> 
> Consider multiplying the timeGap by 10 and last_timeGap by 15 to prevent calling into the floating point libraries. See this: https://github.com/nkapron/Freevalve_Arduino/blob/054797a65dec251340a225ef7532224851ff92c9/valve_control.ino#L142

Actually I like this approach quite a lot! 

@ricallison floats are not "convinient" numbers to work on in computers, they're not exactly what you think they are, also conversion will happen automatically do compare an int to a float, as such losing accuracy. My way of casting the float to int basically turncates the float number to decimals only, i.e. 14.7 => 14 and 14.3 => 14, which was ok since the comparison was for larger, but the multiplication method is better and more accurate, even if potentialy a smige slower because now we have to do 2 multiplications, instead of a casting